### DISCUSSION TITLE
Any and overloads

### DISCUSSION THREAD

**Discussion by mehdigmira** [microsoft/pyright#5922] [Q&A]
In the following example

```python
from typing import Any, Generic, TypeVar, overload, reveal_type


T = TypeVar("T", bound=Any)

class A(Generic[T]):
    ...

@overload
def f(a: A[tuple[int]]) -> tuple[int]:
    ...

@overload
def f(a: A[Any]) -> tuple[Any]:
    ...

def f(a: A[Any]) -> tuple[Any]:
    ...


def test(a: A[Any]):
    reveal_type(f(a))
```

reveal type is tuple[int] which means pyright picks the first overload. I'd say this is an unexpected behaviour, but since Any is a bit of a grey zone typing wise, I'm posting this rather as a discussion than a bug report :)

**Comment by erictraut** [comment]
Pyright's overload behavior is working as intended here. As you said, overload behavior is not well specified. I've spent significant time attempting to make pyright work similarly to mypy in this regard, but even mypy behaves differently from its documented behavior and inconsistently across seemingly-similar examples. I've therefore given up trying to [replicate its behavior exactly](https://microsoft.github.io/pyright/#/mypy-comparison?id=overload-resolution). 

If there is a desire to have 100% consistent overload behavior across type checkers, someone is going to need to specify that behavior in a PEP. All type checkers would then need to agree to modify their current behavior to conform to that spec. I think there would be value in such an effort. If you or someone else is interested in driving an effort to specify this, I would be supportive. I'll note that there is a similar effort underway to formalize other aspects of the type system. Perhaps this could be rolled into that effort.

**Reply by mehdigmira** [reply]
But first overload doesn't get picked up if I change the definition of the first overload to 
```python
from typing import Any, Generic, TypeVar, overload, reveal_type


T = TypeVar("T", bound=Any)

class A(Generic[T]):
    ...

@overload
def f(a: int) -> tuple[int]:
    ...

@overload
def f(a: A[Any]) -> tuple[Any]:
    ...

def f(a: A[Any]) -> tuple[Any]:
    ...


def test(a: A[Any]):
    reveal_type(f(a))
```

The doc you mention also says that it's the last overload that should get picked up. So there smth weird here, no ?

**Reply by erictraut** [reply]
As I said, this is working as intended. Please refer to the documentation if you have questions.

**Reply by mehdigmira** [reply]
I've read the documentation. It says 

> the handling of ambiguous overloads due to Any argument types where one return type is the supertype of all other return types. In this case, pyright evaluates the resulting return type as the supertype

tuple[Any] being the supertype here in both my code snippets, I don't understand why tuple[int] is inferred in the first one. Sorry if I'm missing smth :/

**Reply by erictraut** [reply]
You don't have an `Any` argument type in your example, so the section you quote is not applicable.

The argument type in your code sample is `A[Any]`. This type is clearly not consistent with `int`.

**Reply by mehdigmira** [reply]
ok ! thank you for your time
